src/python/py/models/builder.py

```diff
@@ -1495,6 +1506,128 @@ def make_mlp_fc(self, layer_id, mlp, root_input):
         # Assign output 0 of MLP layer as output of last layer
         self.mlp_attrs["output_0"] = f"{fc2_add_name}/output_0"
 
+    def make_block_sparse_moe(self, layer_id, bsm, root_input):
+        # Make nodes for the MoE subgraph
+        #
+        #                  root_input
+        #                 /       \
+        #         router_MatMul    |
+        #             /     \      |
+        #         Shape      |     |
+        #           |        |     |
+        #         Gather     |     |
+        #           |        |     |
+        #       Unsqueeze    |     |
+        #           |        |    /
+        #        Concat     /    /
+        #             \    /    /
+        #             Reshape  /
+        #                 \   /
+        #                  MoE
+        #                   |
+        #                 output
+        num_experts = self.moe_attrs["num_experts"]
+        top_k = self.moe_attrs["top_k"]
+        activation_type = self.moe_attrs["activation_type"]
+        normalize_routing_weights = self.moe_attrs["normalize_routing_weights"]
+        use_sparse_mixer = self.moe_attrs["use_sparse_mixer"]
+        use_int4 = self.moe_attrs["use_int4"]
+
+        moe_name = f"/model/layers.{layer_id}/moe"
+        gate_ops_base = f"{moe_name}/gate"
+
+        # Make MoE nodes
+        gate_name = f"{gate_ops_base}/MatMul"
+        self.make_matmul(bsm.gate, gate_name, root_input)
+
+        shape_name = f"{gate_ops_base}/Shape"
+        self.make_shape(shape_name, f"{gate_name}/output_0", shape=[3])
+
+        gather_name = f"{gate_ops_base}/Gather"
+        self.make_gather(gather_name, [f"{shape_name}/output_0", "/model/constants/TensorProto.INT64/0D/2"], axis=0)
+
+        unsqueeze_name = f"{gate_ops_base}/Unsqueeze"
+        self.make_unsqueeze(unsqueeze_name, [f"{gather_name}/output_0", "/model/constants/TensorProto.INT64/1D/0"], dtype=TensorProto.INT64, shape=[1])
+
+        concat_name = f"{gate_ops_base}/Concat"
+        self.make_concat(concat_name, ["/model/constants/TensorProto.INT64/1D/-1", f"{unsqueeze_name}/output_0"], dtype=TensorProto.INT64, shape=[2], axis=0)
+
+        gate_reshape_name = f"{gate_ops_base}/Reshape"
+        self.make_reshape(gate_reshape_name, [f"{gate_name}/output_0", f"{concat_name}/output_0"], dtype=self.io_dtype, shape=['num_rows', num_experts])
+
+        def quant_dequant(weights, quant_mode: bool = True):
+            type = torch.quint4x2 if quant_mode else torch.int8
+
+            try:
+                import tensorrt_llm
+
+                _, processed_q_weight, torch_weight_scales = (
+                    torch.ops.trtllm._symmetric_quantize_last_axis_of_batched_matrix(weights.T.cpu().contiguous(), type)
+                )
+            except:
+                print("import tensorrt_llm is needed to use torch.ops.trtllm._symmetric_quantize_last_axis_of_batched_matrix()")
```

Comment:
If `tensorrt_llm` is not installed and it is required to continue, should we raise an `Exception` (or an `ImportError`) instead and add this message there?

Reply:
without trtllm package,  torch.ops.trtllm._symmetric_quantize_last_axis_of_batched_matrix() will throw error that saying .trtllm is not existed in torch. I think we just need an extra message to tell the user how to proceed instead of throwing again.

Reply:
But with the try-except block, that error won't get thrown. Instead it will print this message and continue to this line.

https://github.com/microsoft/onnxruntime-genai/blob/234db941273c66bb9256f802c1648436b87cf364/src/python/py/models/builder.py#L1571

Since `torch_weight_scales = None`, a `NoneType` error will be thrown instead.

Reply:
I just ran it to verify that would be the case.

```
Reading embedding layer
Reading decoder layer 0
import tensorrt_llm is needed to use torch.ops.trtllm._symmetric_quantize_last_axis_of_batched_matrix()
Traceback (most recent call last):
  File "/path/to/onnxruntime-genai/src/python/py/models/builder.py", line 2794, in <module>
    create_model(args.model_name, args.input, args.output, args.precision, args.execution_provider, args.cache_dir, **extra_options)
  File "/path/to/onnxruntime-genai/src/python/py/models/builder.py", line 2686, in create_model
    onnx_model.make_model(input_path)
  File "/path/to/onnxruntime-genai/src/python/py/models/builder.py", line 1768, in make_model
    self.make_layer(self.layer_id, module)
  File "/path/to/onnxruntime-genai/src/python/py/models/builder.py", line 2606, in make_layer
    self.make_block_sparse_moe(layer_id, layer.block_sparse_moe, root_input=self.layernorm_attrs["output_0"])
  File "/path/to/onnxruntime-genai/src/python/py/models/builder.py", line 1582, in make_block_sparse_moe
    w1_scale, pre_qweight1= quant_dequant(bsm.experts[i].w1.weight, use_int4)
  File "/path/to/onnxruntime-genai/src/python/py/models/builder.py", line 1571, in quant_dequant
    return torch_weight_scales.to(torch.float16), processed_q_weight
AttributeError: 'NoneType' object has no attribute 'to'
```

The `AttributeError: 'NoneType' object has no attribute 'to'` message is misleading so I think it would be better to raise an exception and print the `tensorrt_llm` message as part of the exception.

Reply:
you are right, let me change it